### ISSUE TITLE
Teach Travis to Deploy application to container and run integration tests

### ISSUE THREAD

**Issue by BenGalewsky** [OpenTechStrategies/psm#304]
# Problem
As part of PR #279 we need to be able to run the full suite of Serenity tests against a freshly deployed
application, running in the travis container. I've been working on this off and on for several weeks and
the best I can come up with is unreliably unreliable (it works just often enough to make us feel that it should work).

# Configuration
In the `204_selenium_tests` branch the travis.yml file has entries to call a script called `start-wildfly-deploy-application.sh' which:

- Creates the PSM database and user on the travis-provided Postgres server
- Loads the seed data into the database
- Downloads the Postgres jdbc driver
- Starts up wildfly in standalone mode
- Deploy's Postgres driver and creates data sources
- Configures wildlfly to talk to the travis-provided SMTP server
- Finally deploys the freshly build EAR file to the server

The travis.yml file has some additions to it:

- Tells travis to launch a Postgres 9.6 service in the container
- Install Firefox version 46.0.1 (the latest version of Firefox compatible with our selenium drivers)
- Sets environment variables to help selenium find the Firefox binary and X-windows virtual display
- Start the Xwindows virtual display service so we can run Firefox headlessly
- Build the EAR file
- Deploy the EAR file
- Finally run tests against the running application

# Current Situation
There are two common failure modes:

1. When starting wildfly you get an exception that it can't start JMS due to a missing dependency: 
```
19:55:23,058 INFO  [org.jboss.ws.common.management] (MSC service thread 1-6) JBWS022052: Starting JBossWS 5.1.5.Final (Apache CXF 3.1.6)

19:55:23,107 ERROR [org.apache.activemq.artemis.core.server] (ServerService Thread Pool -- 64) AMQ224000: Failure in initialisation: java.lang.RuntimeException: Cannot initialize queue:Resource temporarily unavailable
```

2. After wildfly is reloaded to accept the data sources we deploy the app and we get a different error (which I can't find in my history, but it is similar)

I thought the second failure mode was caused by attempting to deploy in the middle of the restart and tried adding a second call to `wait_for_wildfly` but that introduced a race condition since the restart can happen quickly and that script misses the line in the log file that it is supposed to wait on.

Lately I've seen that I can get past the first failure mode by clearing out the travis cache, build. It fails, try again and it succeeds, try again and it fails thereafter.

**Comment by jasonaowen** [comment]
Last week, @slifty was able to get WildFly to start up on Travis consistently and deploy the newly-built application to it!

He also started working with @cecilia-donnelly to create an account with [Sauce Labs](https://saucelabs.com/open-source) to run our Serenity/Selenium UI integration tests. It turns out there is not much documentation on how our particular combination of tools works, so as of Monday it's a lot of `*make tweak* *push* *wait 5m for travis to tell me it didn't work* *repeat*`.

**Comment by slifty** [comment]
I prodded the keyboard a bit over the past week around integration tests; some progress to report / a summary of the issues faced:

1) It turns out that Sauce Labs needs you to specify that you are creating an account for an open source project during account creation (our free trial ran out, whoops).  I made a personal psm account using {email}+psm@{provider}.com -- I suggest OTS does the same.  They're sneaky about it, you have to manually go to [this URL](saucelabs.com/signup/) since their normal links don't let you pick an open source plan during account creation.

2) After some prodding I got my local build to run tests on Sauce Labs -- one note which might be obvious to everyone else, but you should run ./gradle integration-tests:clean before build; it got stuck on some old output from past experimentation and I thought it was an issue with my settings.  whoops.

3) I had to manually modify the test URL (specifically, the @DefaultUrl in LoginPage.java) to point to our CD server rather than localhost, since sauce runs in the cloud.  I don't know Selenium well enough to know the right way to make this a configurable field; does anybody on the team have that experience?

4) The tests themselves may be out of date; or maybe our master branch has bugs, I'm not sure which, but after running the tests on SauceLabs the results were: 52 tests completed, 6 failed, 9 skipped.  I can post output but the first question is have we been running integration tests before committing to master, or are the tests considered stale?

**Comment by BenGalewsky** [comment]
@slifty - yay! Glad to see this is working. 
According to the [PageObject Documentation](https://github.com/serenity-bdd/serenity-documentation/blob/master/src/asciidoc/page-objects.adoc#opening-the-page) You can override the `@defaultURL` by setting `webdriver.base.url`

**Comment by cecilia-donnelly** [comment]
>It turns out that Sauce Labs needs you to specify that you are creating an account for an open source project during account creation (our free trial ran out, whoops). I made a personal psm account using {email}+psm@{provider}.com -- I suggest OTS does the same. They're sneaky about it, you have to manually go to this URL since their normal links don't let you pick an open source plan during account creation.

I created an open source account for OTS, and will link it to your account.  Thanks for the link to the right place!

>The tests themselves may be out of date; or maybe our master branch has bugs, I'm not sure which, but after running the tests on SauceLabs the results were: 52 tests completed, 6 failed, 9 skipped. I can post output but the first question is have we been running integration tests before committing to master, or are the tests considered stale?

I think @jasonaowen looked at this, and I will do the same.  I don't think we've been systematically running them before committing to master.

Thanks, @slifty!

**Comment by jasonaowen** [comment]
> The tests themselves may be out of date; or maybe our master branch has bugs, I'm not sure which, but after running the tests on SauceLabs the results were: 52 tests completed, 6 failed, 9 skipped. I can post output but the first question is have we been running integration tests before committing to master, or are the tests considered stale?

We have been running the integration tests inconsistently, but a lot recently as I was splitting up/rewriting #332 and getting it merged. I was able to run them successfully just now.

Which are failing, @slifty?

**Comment by cecilia-donnelly** [comment]
Hurray! @slifty merged #483, which fixed this.  We now have the integration tests running as part of our CI process.
